c7n/filters/core.py

```diff
@@ -971,3 +974,64 @@ def reorder(self, items, key=None):
             return items[::-1]
         else:
             return sorted(items, key=key, reverse=(self.order == 'desc'))
+
+
+class CELFilter(Filter):
+    """Generic CEL filter using CELPY"""
+
+    def __init__(self, data, manager):
+        super().__init__(data, manager)
+        assert data["type"].lower() == "cel"
+        # throw error if expr missing? or run and just return resources as if empty filter applied?
+        self.expr = data.get("expr")
+        self.parser = None
+        self.cel_env = None
+        self.cel_ast = None
+        self.decls = {
+            "Resource": celpy.celtypes.MapType,
+            "Now": celpy.celtypes.TimestampType
+        }
+        self.decls.update(celpy.c7nlib.DECLARATIONS)
+
+    schema = {
+        'type': 'object',
+        'additionalProperties': False,
+        'required': ['type'],
+        'properties': {
+            'type': {'enum': ['cel']},
+            'expr': {'type': 'string'}
+        }
+    }
+    schema_alias = True
+    annotate = True
+    required_keys = {'cel', 'expr'}
+
+    def validate(self):
+        for filter in self.manager.data["filters"]:
+            if 'expr' not in filter:
+                raise PolicyValidationError(
+                    f"CEL filters can only be used with provided expressions in {self.manager.data}"
+                )
+
+        self.cel_env = celpy.Environment(
+            annotations=self.decls,
+            runner_class=celpy.c7nlib.C7N_Interpreted_Runner
+        )
+        self.cel_ast = self.cel_env.compile(self.data["expr"])
+        return self
+
+    def process(self, resources, event=None, filter=Filter):
+        filtered_resources = []
+        for resource in resources:
+            cel_prgm = self.cel_env.program(self.cel_ast, functions=celpy.c7nlib.FUNCTIONS)
+            cel_activation = {
+                "Resource": celpy.json_to_cel(resource),
```

Comment:
and let's lower case the activate var names, and simplify Resource to `r` the goal is multi attribute matching and the duplication in that case seems redundant.. doc string example as well.

Reply:
Got it, working on this now but dependent on https://github.com/cloud-custodian/cel-python/pull/19